File: src/rs/cache/format/Container.ts
// import { Xtea } from "../util/Xtea";
import { CompressionHandler } from "../../compression/CompressionHandler";
import { CompressionType } from "../../compression/CompressionType";
import { Xtea, XteaKey } from "../../crypto/Xtea";
import { ByteSource } from "../../io/ByteSource";
import { ByteSourceReader } from "../../io/ByteSourceReader";
import { getOrCopyBytes } from "../../io/ByteSourceUtil";
import { readU32BE } from "../../io/Endian";

export class Container {
    static decodeFromSource(
        source: ByteSource,
        key: XteaKey | null,
        compressionHandler: CompressionHandler,
    ): Container {
        const reader = new ByteSourceReader(source);
        const compression: CompressionType = reader.readUnsignedByte() as CompressionType;
        const size = reader.readInt();

        if (size < 0) {
            throw new Error("Invalid container size: " + size);
        }

        const hasKey = Xtea.isValidKey(key);

        if (compression === CompressionType.None) {
            if (hasKey) {
                const encryptedSize = 4 + size;
                const end = 5 + encryptedSize;
                if (end > source.size) {
                    throw new Error(
                        `Truncated container. expected>=${end}, available=${source.size}`,
                    );
                }

                const encrypted = new Uint8Array(encryptedSize);
                source.readInto(5, encrypted);
                Xtea.decryptInPlace(encrypted, 0, encryptedSize, key);

                return new Container(compression, encrypted.subarray(0, size));
            }

            const end = 5 + size;
            if (end > source.size) {
                throw new Error(`Truncated container. expected>=${end}, available=${source.size}`);
            }

            const payload = source.slice(5, size);
            return new Container(compression, getOrCopyBytes(payload));
        }

        if (compression !== CompressionType.Bzip2 && compression !== CompressionType.Gzip) {
            throw new Error("Container: Unsupported compression: " + compression);
        }

        const compressedSize = size;
        const expectedMinSize = 5 + 4 + compressedSize;
        if (expectedMinSize > source.size) {
            throw new Error(
                `Truncated container. expected>=${expectedMinSize}, available=${source.size}`,
            );
        }

        let actualSize: number;
        let compressed: Uint8Array;

        if (hasKey) {
            const encryptedSize = 4 + compressedSize;
            const encrypted = new Uint8Array(encryptedSize);
            source.readInto(5, encrypted);
            Xtea.decryptInPlace(encrypted, 0, encryptedSize, key);

            actualSize = readU32BE(encrypted, 0);
            compressed = encrypted.subarray(4, 4 + compressedSize);
        } else {
            reader.seek(5);
            actualSize = reader.readUnsignedInt();

            // compressed payload starts at 9
            const payload = source.slice(9, compressedSize);
            compressed = getOrCopyBytes(payload);
        }

        let decompressed: Uint8Array;
        if (compression === CompressionType.Bzip2) {
            decompressed = compressionHandler.decompressBzip2(compressed, actualSize);
        } else {
            decompressed = compressionHandler.decompressGzip(compressed);
        }

        if (decompressed.length !== actualSize) {
            throw new Error(
                "Container: Size mismatch. Compressed: " +
                    actualSize +
                    ", Decompressed: " +
                    decompressed.length +
                    ", Type: " +
                    CompressionType[compression],
            );
        }

        return new Container(compression, decompressed);
    }

    constructor(
        readonly compression: CompressionType,
        readonly data: Uint8Array,
    ) {}
}

File: cpp/rs/cache/format/Container.cpp
#include "Container.hpp"

#include <cstddef>

#include "../../compression/CompressionHandler.hpp"
#include "../../compression/CompressionType.hpp"
#include "../../core/Allocator.hpp"
#include "../../core/Move.hpp"
#include "../../core/Result.hpp"
#include "../../core/Span.hpp"
#include "../../core/Status.hpp"
#include "../../core/Vec.hpp"
#include "../../crypto/Xtea.hpp"
#include "../../io/ByteSource.hpp"
#include "../../io/Endian.hpp"
#include "../../types.hpp"

namespace rs {

Result<Container> Container::decodeFromSource(
    const ByteSource& source,
    const XteaKey* key,
    const CompressionHandler& compressionHandler,
    Allocator& alloc) noexcept {
    if (source.size() < 5) {
        return Result<Container>::err(Status::Truncated);
    }

    u8 head[5];
    Status s = source.readInto(0, Span<u8>(head, sizeof(head)));
    if (!ok(s)) {
        return Result<Container>::err(s);
    }

    const CompressionType compression = static_cast<CompressionType>(head[0]);
    const i32 sizeSigned = readI32BE(head + 1);
    if (sizeSigned < 0) {
        return Result<Container>::err(Status::BadFormat);
    }
    const std::size_t size = static_cast<std::size_t>(sizeSigned);

    const bool hasKey = Xtea::isValidKey(key);

    if (compression == CompressionType::None) {
        if (hasKey) {
            const std::size_t encryptedSize = 4 + size;
            const std::size_t end = 5 + encryptedSize;
            if (end > source.size()) {
                return Result<Container>::err(Status::Truncated);
            }

            Vec<u8> encrypted(alloc);
            auto rr = encrypted.resize(encryptedSize);
            if (!rr.isOk()) {
                return Result<Container>::err(rr.status());
            }
            s = source.readInto(5, Span<u8>(encrypted.data(), encrypted.size()));
            if (!ok(s)) {
                return Result<Container>::err(s);
            }
            s = Xtea::decryptInPlace(Span<u8>(encrypted.data(), encrypted.size()), 0, encryptedSize, *key);
            if (!ok(s)) {
                return Result<Container>::err(s);
            }

            rr = encrypted.resize(size);
            if (!rr.isOk()) {
                return Result<Container>::err(rr.status());
            }
            return Result<Container>::ok(Container(compression, rs::move(encrypted)));
        }

        const std::size_t end = 5 + size;
        if (end > source.size()) {
            return Result<Container>::err(Status::Truncated);
        }

        Vec<u8> out(alloc);
        auto rr = out.resize(size);
        if (!rr.isOk()) {
            return Result<Container>::err(rr.status());
        }
        if (size) {
            s = source.readInto(5, Span<u8>(out.data(), out.size()));
            if (!ok(s)) {
                return Result<Container>::err(s);
            }
        }
        return Result<Container>::ok(Container(compression, rs::move(out)));
    }

    if (compression != CompressionType::Bzip2 && compression != CompressionType::Gzip) {
        return Result<Container>::err(Status::Unsupported);
    }

    const std::size_t compressedSize = size;
    const std::size_t expectedMinSize = 5 + 4 + compressedSize;
    if (expectedMinSize > source.size()) {
        return Result<Container>::err(Status::Truncated);
    }

    u32 actualSize = 0;
    Vec<u8> compressed(alloc);
    Span<const u8> compressedView;

    if (hasKey) {
        const std::size_t encryptedSize = 4 + compressedSize;
        Vec<u8> encrypted(alloc);
        auto rr = encrypted.resize(encryptedSize);
        if (!rr.isOk()) {
            return Result<Container>::err(rr.status());
        }
        s = source.readInto(5, Span<u8>(encrypted.data(), encrypted.size()));
        if (!ok(s)) {
            return Result<Container>::err(s);
        }
        s = Xtea::decryptInPlace(Span<u8>(encrypted.data(), encrypted.size()), 0, encryptedSize, *key);
        if (!ok(s)) {
            return Result<Container>::err(s);
        }

        actualSize = readU32BE(encrypted.data());
        compressedView = Span<const u8>(encrypted.data() + 4, compressedSize);

        Result<Vec<u8>> decRes = (compression == CompressionType::Bzip2)
            ? compressionHandler.decompressBzip2(compressedView, static_cast<std::size_t>(actualSize), alloc)
            : compressionHandler.decompressGzip(compressedView, alloc);

        if (!decRes.isOk()) {
            return Result<Container>::err(decRes.status());
        }
        Vec<u8> decompressed = rs::move(decRes.value());
        if (decompressed.size() != static_cast<std::size_t>(actualSize)) {
            return Result<Container>::err(Status::SizeMismatch);
        }
        return Result<Container>::ok(Container(compression, rs::move(decompressed)));
    }

    u8 actualBuf[4];
    s = source.readInto(5, Span<u8>(actualBuf, sizeof(actualBuf)));
    if (!ok(s)) {
        return Result<Container>::err(s);
    }
    actualSize = readU32BE(actualBuf);

    {
        auto rr = compressed.resize(compressedSize);
        if (!rr.isOk()) {
            return Result<Container>::err(rr.status());
        }
        if (compressedSize) {
            s = source.readInto(9, Span<u8>(compressed.data(), compressed.size()));
            if (!ok(s)) {
                return Result<Container>::err(s);
            }
        }
        compressedView = Span<const u8>(compressed.data(), compressed.size());
    }

    Result<Vec<u8>> decRes = (compression == CompressionType::Bzip2)
        ? compressionHandler.decompressBzip2(compressedView, static_cast<std::size_t>(actualSize), alloc)
        : compressionHandler.decompressGzip(compressedView, alloc);

    if (!decRes.isOk()) {
        return Result<Container>::err(decRes.status());
    }
    Vec<u8> decompressed = rs::move(decRes.value());
    if (decompressed.size() != static_cast<std::size_t>(actualSize)) {
        return Result<Container>::err(Status::SizeMismatch);
    }
    return Result<Container>::ok(Container(compression, rs::move(decompressed)));
}

} // namespace rs
